driver-core/src/main/java/com/datastax/driver/core/DataTypeParser.java

```diff
@@ -0,0 +1,255 @@
+            .build();
+
+    static DataType parse(String toParse, Cluster cluster, String currentKeyspaceName, Map<String, UserType> userTypes, boolean frozen) {
+
+        Parser parser = new Parser(toParse, 0);
+        String type = parser.parseTypeName();
+
+        DataType nativeType = NATIVE_TYPES_MAP.get(type.toLowerCase());
+        if (nativeType != null)
+            return nativeType;
+
+        if (type.equalsIgnoreCase(LIST)) {
+            List<String> parameters = parser.parseTypeParameters();
+            if (parameters.size() != 1)
+                throw new DriverInternalError(String.format("Excepting single parameter for list, got %s", parameters));
+            DataType elementType = parse(parameters.get(0), cluster, currentKeyspaceName, userTypes, false);
+            return list(elementType, frozen);
+        }
+
+        if (type.equalsIgnoreCase(SET)) {
+            List<String> parameters = parser.parseTypeParameters();
+            if (parameters.size() != 1)
+                throw new DriverInternalError(String.format("Excepting single parameter for set, got %s", parameters));
+            DataType elementType = parse(parameters.get(0), cluster, currentKeyspaceName, userTypes, false);
+            return set(elementType, frozen);
+        }
+
+        if (type.equalsIgnoreCase(MAP)) {
+            List<String> parameters = parser.parseTypeParameters();
+            if (parameters.size() != 2)
+                throw new DriverInternalError(String.format("Excepting two parameters for map, got %s", parameters));
+            DataType keyType = parse(parameters.get(0), cluster, currentKeyspaceName, userTypes, false);
+            DataType valueType = parse(parameters.get(1), cluster, currentKeyspaceName, userTypes, false);
+            return map(keyType, valueType, frozen);
+        }
+
+        if (type.equalsIgnoreCase(FROZEN)) {
+            List<String> parameters = parser.parseTypeParameters();
+            if (parameters.size() != 1)
+                throw new DriverInternalError(String.format("Excepting single parameter for frozen keyword, got %s", parameters));
+            return parse(parameters.get(0), cluster, currentKeyspaceName, userTypes, true);
+        }
+
+        if (type.equalsIgnoreCase(TUPLE)) {
+            List<String> rawTypes = parser.parseTypeParameters();
+            List<DataType> types = new ArrayList<DataType>(rawTypes.size());
+            for (String rawType : rawTypes) {
+                types.add(parse(rawType, cluster, currentKeyspaceName, userTypes, false));
+            }
+            return cluster.getMetadata().newTupleType(types);
+        }
+
+        // return a custom type for the special empty type
+        // so that it gets detected later on, see TableMetadata
+        if(type.equalsIgnoreCase(EMPTY))
+            return custom(type);
+
+        // Try UDTs
+        if (userTypes != null) {
+            UserType userType = userTypes.get(Metadata.handleId(type));
```

Comment:
I was surprised by the call to `handleId` here, but Cassandra does indeed quote mixed-case names in its type columns:
```
cqlsh:test> create type "MixedCaseType"(i int);
cqlsh:test> create table foo(k int primary key, t frozen<"MixedCaseType">);
cqlsh:test> select type from system_schema.columns where keyspace_name = 'test' and table_name = 'foo' and column_name = 't';

 type
-------------------------
 frozen<"MixedCaseType">
```